### DISCUSSION TITLE
Can't url encode chinese characters

### DISCUSSION THREAD

**Discussion by GlassGruber** [curl/curl#10446] [Q&A] [answered]
I'm facing a weird bug, but I fear this is more of a RTFM problem so please bear with me.

The last version of curl for windows 64bit is not url encoding this chinese example characters, and instead only question marks are sent `????`. I've tested with many more characters and the result is the same. 

Here below the command to reproduce my issue, I've placed also the relevant portion of the trace dump. 
I would expect the encoded part as this `%E7%A7%8B%E6%94%B6%E5%86%AC%E8%97%8F%E5%A5%B3`

What am I doing wrong? 
Thank you!

```
curl.exe --trace - 'https://httpbin.org/anything' -X POST --data-urlencode "text=秋收冬藏女" 
```

```
=> Send data, 20 bytes (0x14)
0000: 74 65 78 74 3d 25 33 46 25 33 46 25 33 46 25 33 text=%3F%3F%3F%3
0010: 46 25 33 46                                     F%3F
...
<= Recv data, 452 bytes (0x1c4)
0000: 7b 0a 20 20 22 61 72 67 73 22 3a 20 7b 7d 2c 20 {.  "args": {},
0010: 0a 20 20 22 64 61 74 61 22 3a 20 22 22 2c 20 0a .  "data": "", .
0020: 20 20 22 66 69 6c 65 73 22 3a 20 7b 7d 2c 20 0a   "files": {}, .
0030: 20 20 22 66 6f 72 6d 22 3a 20 7b 0a 20 20 20 20   "form": {.
0040: 22 74 65 78 74 22 3a 20 22 3f 3f 3f 3f 3f 22 0a "text": "?????".
0050: 20 20 7d 2c 20 0a 20 20 22 68 65 61 64 65 72 73   }, .  "headers
0060: 22 3a 20 7b 0a 20 20 20 20 22 41 63 63 65 70 74 ": {.    "Accept
0070: 22 3a 20 22 2a 2f 2a 22 2c 20 0a 20 20 20 20 22 ": "*/*", .    "
0080: 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 22 3a Content-Length":
0090: 20 22 32 30 22 2c 20 0a 20 20 20 20 22 43 6f 6e  "20", .    "Con
00a0: 74 65 6e 74 2d 54 79 70 65 22 3a 20 22 61 70 70 tent-Type": "app
00b0: 6c 69 63 61 74 69 6f 6e 2f 78 2d 77 77 77 2d 66 lication/x-www-f
00c0: 6f 72 6d 2d 75 72 6c 65 6e 63 6f 64 65 64 22 2c orm-urlencoded",
00d0: 20 0a 20 20 20 20 22 48 6f 73 74 22 3a 20 22 68  .    "Host": "h
00e0: 74 74 70 62 69 6e 2e 6f 72 67 22 2c 20 0a 20 20 ttpbin.org", .
00f0: 20 20 22 55 73 65 72 2d 41 67 65 6e 74 22 3a 20   "User-Agent":
0100: 22 63 75 72 6c 2f 37 2e 38 37 2e 30 22 2c 20 0a "curl/7.87.0", .
0110: 20 20 20 20 22 58 2d 41 6d 7a 6e 2d 54 72 61 63     "X-Amzn-Trac
0120: 65 2d 49 64 22 3a 20 22 52 6f 6f 74 3d 31 2d 36 e-Id": "Root=1-6
0130: 33 65 34 30 39 35 36 2d 34 37 30 35 32 30 39 61 3e40956-4705209a
0140: 33 34 34 31 36 63 63 38 33 37 35 38 64 38 31 64 34416cc83758d81d
0150: 22 0a 20 20 7d 2c 20 0a 20 20 22 6a 73 6f 6e 22 ".  }, .  "json"
0160: 3a 20 6e 75 6c 6c 2c 20 0a 20 20 22 6d 65 74 68 : null, .  "meth
0170: 6f 64 22 3a 20 22 50 4f 53 54 22 2c 20 0a 20 20 od": "POST", .
0180: 22 6f 72 69 67 69 6e 22 3a 20 22 32 2e 33 36 2e "origin": "2.36.
0190: 31 39 35 2e 36 31 22 2c 20 0a 20 20 22 75 72 6c 195.61", .  "url
01a0: 22 3a 20 22 68 74 74 70 73 3a 2f 2f 68 74 74 70 ": "https://http
01b0: 62 69 6e 2e 6f 72 67 2f 61 6e 79 74 68 69 6e 67 bin.org/anything
01c0: 22 0a 7d 0a                                     ".}.
{
  "args": {},
  "data": "",
  "files": {},
  "form": {
    "text": "?????"
  },
  "headers": {
    "Accept": "*/*",
    "Content-Length": "20",
    "Content-Type": "application/x-www-form-urlencoded",
    "Host": "httpbin.org",
    "User-Agent": "curl/7.87.0",
    "X-Amzn-Trace-Id": "Root=1-63e40956-4705209a34416cc83758d81d"
  },
  "json": null,
  "method": "POST",
  "origin": "2.36.195.61",
  "url": "https://httpbin.org/anything"
}
== Info: [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS header, Supplemental data (23):
<= Recv SSL data, 5 bytes (0x5)
0000: 17 03 03 00 21                                  ....!
== Info: Connection #0 to host httpbin.org left intact
```

**Comment by jay** [comment]
What is your `curl -V` info

**Reply by bagder** [reply]
Also, where/how do you invoke that curl command line? Is that a regular "cmd" shell or something else?

**Reply by GlassGruber** [reply]
Hei there!
Very sorry should have added from start, this is the `-V` output, I've downloaded the Windows bin from curl website.

```
curl 7.87.0 (x86_64-w64-mingw32) libcurl/7.87.0 OpenSSL/3.0.8 (Schannel) zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 WinIDN libssh2/1.10.0 nghttp2/1.51.0 ngtcp2/0.13.0 nghttp3/0.8.0 libgsasl/2.2.0
Release-Date: 2022-12-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe TLS-SRP UnixSockets zstd
```

> Also, where/how do you invoke that curl command line? Is that a regular "cmd" shell or something else?

Interesting, I've tested both on CMD and Git Bash in Cmder, but also tried with "vanilla" CMD without any intermediary terminal and similar.

Thank you both for your time!

**Reply by GlassGruber** [reply]
Out of curiosity I've tested also with Powershell Core and got same results.

**Comment by jay** [comment, marked as answer]
In Windows if curl is not built with Unicode support (no Unicode in `Features` version info) then the command line text is treated in the current locale instead of UTF-8. The [official build of curl for Windows](https://curl.se/windows/) (what you are using) does not have Unicode support so say your locale is Chinese (codepage 936) then `text=秋收冬藏女` chinese characters is 5 multibyte (DBCS) two bytes each:
~~~
=> Send data, 35 bytes (0x23)
0000: 74 65 78 74 3d 25 43 37 25 45 46 25 43 41 25 44 text=%C7%EF%CA%D
0010: 35 25 42 36 25 41 43 25 42 32 25 44 38 25 43 35 5%B6%AC%B2%D8%C5
0020: 25 41 45                                        %AE
~~~

Since httpbin is expecting to echo back Unicode characters it won't understand that and return bad Unicode:

~~~
  "form": {
    "text": "\ufffd\ufffd\ufffd\u0576\ufffd\ufffd\ufffd\u016e"
  },
~~~

OTOH curl built with Unicode support will read and send the command line as Unicode characters in UTF-8 encoding:

~~~
=> Send data, 50 bytes (0x32)
0000: 74 65 78 74 3d 25 45 37 25 41 37 25 38 42 25 45 text=%E7%A7%8B%E
0010: 36 25 39 34 25 42 36 25 45 35 25 38 36 25 41 43 6%94%B6%E5%86%AC
0020: 25 45 38 25 39 37 25 38 46 25 45 35 25 41 35 25 %E8%97%8F%E5%A5%
0030: 42 33                                           B3
~~~

that results in expected server output (ie same Unicode characters in are returned):

~~~
  "form": {
    "text": "\u79cb\u6536\u51ac\u85cf\u5973"
  },
~~~

In this case you will need a build with Unicode support or make a UTF-8 encoded text file with content `秋收冬藏女` and have curl read from the file, eg `--data-urlencode text@filename` which would send `text=%E7%A7%8B%E6%94%B6%E5%86%AC%E8%97%8F%E5%A5%B3` in builds with and without Unicode.

**Reply by GlassGruber** [reply]
Thanks @jay for your thorough explanation and your time!

The file read suggestion is very useful thank you! Is this UTF-8 encoding mentioned somewhere in the docs?
Anyway this totally make sense, yet a bit unexpected for me, what is the reasoning for not building with UTF-8 support? 
I'm genuinely asking, are there some technical background issues, or maybe is there some thread where this has been discussed?
For more context I'm on US Locale and codepage 65001 and the output is filled with question marks.

In conclusion I've worked around the problem by using JQ to url encode the text with something like this (in case someone may ever need it)

```bash
curl.exe 'https://httpbin.org/anything' -X POST --data-raw "text=$(echo -n 秋收冬藏女 | jq -sRj @uri)"
```

but the file reading approach is definetly the way to go to avoid external dependencies and can be simplified as this

```bash
echo -n 秋收冬藏女 | curl.exe 'https://httpbin.org/anything' -X POST  --data-urlencode text@-
```

Note the dash with `text@-`, which is not stated explicitly [in the docs](https://everything.curl.dev/http/post/url-encode#:~:text=in%20the%20POST.-,name%40filename,-%3A%20This%20will%20make), that will allow to read from STDIN instead of a file.

**Reply by jay** [reply]
> Anyway this totally make sense, yet a bit unexpected for me, what is the reasoning for not building with UTF-8 support?

There was some encoding or compatibility difference that the maintainer of the official curl build for Windows did not like. Basically Windows did not support UTF-8 as a locale, or did not support it properly, so in Unicode builds curl continues to run in the user's current locale and then some strings are expected to be UTF-8 encoded either by the user or by the curl tool converting the command line arguments to UTF-8.

Microsoft CRT [added](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-140#utf-8-support) support for UTF-8 locale in Windows 10, so conceivably changing the locale to UTF-8 in Unicode builds may improve the situation. I don't remember if we already tried this or not. 

/cc @vszakats

**Reply by vszakats** [reply]
Well, this isn't about what "I like".

Unicode has been enabled for a year in official builds, right after its state seemed mature enough for an experiment. Turns out that similar reports kept coming, though with expecting the reverse outcome (even though some of those expected outcomes were just so by lucky chance due to how non-Unicode builds worked before). Plus reports about valid Unicode build problems.

Then I made an attempt to upgrade a libcurl API binding for Unicode and had to face the problem also described by OP: There is no documentation for the expected encoding/codepage of input/output strings. The reason for this is simple: This isn't defined in the libcurl API. libcurl API deals with raw byte strings and that's all. This works fine with sane systems like mostly anything besides Windows. But the Windows API and its CRT speak UTF-16 in Unicode mode, which requires conversion. For that conversion we must know what to convert from/to and for which strings exactly. Not knowing that, it was not possible to upgrade the bindings. And with Unicode enabled, the behaviour was even less predictable than before.

Then I made this wiki page in an attempt to plan/document the expected encodings: https://github.com/curl/curl/wiki/libcurl-and-expected-string-encodings
This did not catch on.

Meanwhile a Unicode issue was addressed by Jay in a PR. In that PR, the concept of auto-detecting the input encoding (= UTF-8 or not?) was also introduced, mashed together with other fixes. I "did not like" this indeed, not due a personal opinion, but because auto-detecting UTF-8 is a _heuristic_ method, which works some of the time, but not every time. Introducing it might make some reported issues disappear, but would introduce new and more obscure ones.

IMO to address these, curl (tool and API) would need to give some control over the string encodings. E.g. by adding a way to pass a raw byte string via the command-line in hex representation. Add a flag/setting to tell if the input encoding is UTF-8 or not, and there is probably a bunch of API boundaries where the encoding isn't translated correctly, or at all. (And places where it is translated, while it shouldn't be.)

Some dependencies/interfaces also may or may not support Unicode at all, or with quirks. Also, if the wire protocol supports it, curl would need to send/accept the necessary encoding-related information. (E.g. curl ATM doesn't support a Unicode 'filename*=UTF-8' when using `--remote-header-name`.)

While the original hope was that Windows Unicode support will continue to improve, this did not happen.

More details and links in the commit below:
https://github.com/curl/curl-for-win/commit/bb029053639336c64d103fc97c8e33a495df20f1

UPDATE: Added a new `unicode` label and tagged pending and some past issues with it.

**Reply by GlassGruber** [reply]
Very interesting, and unfortunately I kinda suspected that in the end this was all a Windows compatibility issue! What I really didn't know is that after all this time UTF-8 is still not set in stone as "problem solved" for Windows.

Thank you both for time and sharing this background knowledge!

**Reply by vszakats** [reply]
Bumped into this MS article from April 2022:
https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page

It says that the Windows "ANSI" (non-Unicode) API can now operate on UTF-8 strings (requires Windows 10 v1903+). Conditions apply, but it may eventually help fixing this in curl.

UPDATE 1: A real-world issue (the request part) successfully solved by enabling the above Windows feature, using a non-Unicode `curl.exe`: https://github.com/curl/curl/issues/10820#issue-1636689996

UPDATE 2: Ref: https://mastodon.gamedev.place/@AshleyGullen/111109299141510319
